File: java/Baseline.java
import io.pmem.pmemkv.KVEngine;

import java.io.File;
import java.util.concurrent.atomic.AtomicInteger;

public class Baseline {

    public final static int COUNT = 1000000;
    public final static String PATH = "/dev/shm/pmemkv";

    public static void test_engine(String engine, byte[][] keys, byte[] value) {
        System.out.printf("%nTesting %s engine for %s keys, value size is %s...%n", engine, COUNT, value.length);
        (new File(PATH)).delete();
        KVEngine kv = new KVEngine(engine, "{\"path\":\"" + PATH + "\"}");

        System.out.printf("Put (sequential series)%n");
        long start = System.currentTimeMillis();
        for (byte[] k : keys) kv.put(k, value);
        double elapsed = (System.currentTimeMillis() - start) / 1000.0;
        System.out.printf("  in %f sec%n", elapsed);

        System.out.printf("Get (sequential series)%n");
        int failures = 0;
        start = System.currentTimeMillis();
        for (byte[] k : keys) {
            byte[] result = kv.get(k);
            if (result == null || result.length == 0) failures++;
        }
        elapsed = (System.currentTimeMillis() - start) / 1000.0;
        System.out.printf("  in %f sec, failures=%d%n", elapsed, failures);

        System.out.printf("Exists (sequential series)%n");
        failures = 0;
        start = System.currentTimeMillis();
        for (byte[] k : keys) if (!kv.exists(k)) failures++;
        elapsed = (System.currentTimeMillis() - start) / 1000.0;
        System.out.printf("  in %f sec, failures=%d%n", elapsed, failures);

        System.out.printf("All (one pass)%n");
        AtomicInteger callbacks = new AtomicInteger(0);
        start = System.currentTimeMillis();
        kv.all((byte[] k) -> callbacks.incrementAndGet());
        elapsed = (System.currentTimeMillis() - start) / 1000.0;
        System.out.printf("  in %f sec, failures=%d%n", elapsed, keys.length - callbacks.get());

        System.out.printf("Each (one pass)%n");
        callbacks.set(0);
        start = System.currentTimeMillis();
        kv.each((byte[] k, byte[] v) -> callbacks.incrementAndGet());
        elapsed = (System.currentTimeMillis() - start) / 1000.0;
        System.out.printf("  in %f sec, failures=%d%n", elapsed, keys.length - callbacks.get());

        kv.stop();
    }

    public static void main(String[] args) {
        // format all keys in advance
        byte[][] keys = new byte[COUNT][];
        for (int i = 0; i < COUNT; i++) {
            keys[i] = Integer.toString(i).getBytes();
        }

        // test all engines for all keys & values
        test_engine("blackhole", keys, "AAAAAAAAAAAAAAAA".getBytes());
        test_engine("tree3", keys, "AAAAAAAAAAAAAAAA".getBytes());
        test_engine("tree3", keys, (new String(new char[200]).replace("\0", "A")).getBytes());
        test_engine("tree3", keys, (new String(new char[800]).replace("\0", "A")).getBytes());

        System.out.printf("%nFinished!%n%n");
    }

}

File: nodejs/baseline.js
const fs = require('fs');
const KVEngine = require('../../pmemkv-nodejs/lib/kv_engine');

const COUNT = 1000000;
const PATH = '/dev/shm/pmemkv';

function test_engine(engine, value) {
    console.log(`\nTesting ${engine} engine for ${COUNT} keys, value size is ${value.length}...`);
    if (fs.existsSync(PATH)) fs.unlinkSync(PATH);
    const kv = new KVEngine(engine, `{"path":"${PATH}"}`);

    console.log(`Put (sequential series)`);
    console.time("  in");
    for (let i = 1; i <= COUNT; i++) {
        kv.put(i.toString(), value);
    }
    console.timeEnd("  in");

    console.log(`Get (sequential series)`);
    let failures = 0;
    console.time("  in");
    for (let i = 1; i <= COUNT; i++) {
        const result = kv.get(i.toString());
        if (typeof result === 'undefined' || result === null) failures++;
    }
    console.timeEnd("  in");
    console.log(`  failures: ${failures}`);

    console.log(`Exists (sequential series)`);
    failures = 0;
    console.time("  in");
    for (let i = 1; i <= COUNT; i++) {
        if (!kv.exists(i.toString())) failures++;
    }
    console.timeEnd("  in");
    console.log(`  failures: ${failures}`);

    console.log(`All (one pass)`);
    failures = COUNT;
    console.time("  in");
    kv.all(() => failures--);
    console.timeEnd("  in");
    console.log(`  failures: ${failures}`);

    console.log(`Each (one pass)`);
    failures = COUNT;
    console.time("  in");
    kv.each(() => failures--);
    console.timeEnd("  in");
    console.log(`  failures: ${failures}`);

    kv.stop();
}

// test all engines for all keys & values
test_engine('blackhole', 'AAAAAAAAAAAAAAAA');
test_engine('tree3', 'AAAAAAAAAAAAAAAA');
test_engine('tree3', 'A'.repeat(200));
test_engine('tree3', 'A'.repeat(800));

console.log('\nFinished!\n\n');
